crates/bevy_ecs/src/event.rs

```diff
@@ -798,48 +810,77 @@ impl<'a, E: Event> ExactSizeIterator for EventIteratorWithId<'a, E> {
+    // occur.
+    event_updates: Vec<(ComponentId, unsafe fn(MutUntyped))>,
+}
+
+impl EventRegistry {
+    /// Registers an event type to be updated.
+    pub fn register_event<T: Event>(world: &mut World) {
+        // By initializing the resource here, we can be sure that it is present,
+        // and receive the correct, up-to-date `ComponentId` even if it was previously removed.
+        let component_id = world.init_resource::<Events<T>>();
+        let mut registry = world.get_resource_or_insert_with(Self::default);
+        registry.event_updates.push((component_id, |ptr| {
+            // SAFETY: The resource was initialized with the type Events<T>.
+            unsafe { ptr.with_type::<Events<T>>() }
+                .bypass_change_detection()
+                .update();
+        }));
+    }
+
+    /// Updates all of the registered events in the World.
+    pub fn run_updates(&self, world: &mut World, last_change_tick: Tick) {
+        for (component_id, update) in &self.event_updates {
+            // Bypass the type ID -> Component ID lookup with the cached component ID.
+            if let Some(events) = world.get_resource_mut_by_id(*component_id) {
+                if events.has_changed_since(last_change_tick) {
+                    // SAFETY: The update function pointer is called with the resource
+                    // fetched from the same component ID.
+                    unsafe { update(events) };
+                }
+            }
+        }
+    }
+}
 
 #[doc(hidden)]
 #[derive(SystemSet, Clone, Debug, PartialEq, Eq, Hash)]
 pub struct EventUpdates;
 
 /// Signals the [`event_update_system`] to run after `FixedUpdate` systems.
-pub fn signal_event_update_system(signal: Option<ResMut<EventUpdateSignal>>) {
-    if let Some(mut s) = signal {
-        s.0 = true;
-    }
-}
-
-/// Resets the `EventUpdateSignal`
-pub fn reset_event_update_signal_system(signal: Option<ResMut<EventUpdateSignal>>) {
-    if let Some(mut s) = signal {
-        s.0 = false;
+pub fn signal_event_update_system(signal: Option<ResMut<EventRegistry>>) {
+    if let Some(mut registry) = signal {
+        registry.needs_update = true;
     }
 }
 
-/// A system that calls [`Events::update`].
-pub fn event_update_system<T: Event>(
-    update_signal: Option<Res<EventUpdateSignal>>,
-    mut events: ResMut<Events<T>>,
-) {
-    if let Some(signal) = update_signal {
-        // If we haven't got a signal to update the events, but we *could* get such a signal
-        // return early and update the events later.
-        if !signal.0 {
-            return;
-        }
+/// A system that calls [`Events::update`] on all registered [`Events`] in the world.
+pub fn event_update_system(world: &mut World, mut last_change_tick: Local<Option<Tick>>) {
+    if world.contains_resource::<EventRegistry>() {
+        let change_tick = *last_change_tick;
+        let change_tick = change_tick.unwrap_or(Tick::new(0));
```

Comment:
We could impl `Default` for `Tick` and avoid the Option dance here. I think that's a pretty reasonable idea (especially if we swap to u64 ticks). Doesn't need to be this PR though.

Reply:
Done.